warehouse/query-core/src/main/java/datawave/next/BaseDocIdIterator.java

```diff
@@ -0,0 +1,284 @@
+package datawave.next;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.TreeSet;
+
+import org.apache.accumulo.core.data.ArrayByteSequence;
+import org.apache.accumulo.core.data.ByteSequence;
+import org.apache.accumulo.core.data.Key;
+import org.apache.accumulo.core.data.Range;
+import org.apache.accumulo.core.data.Value;
+import org.apache.accumulo.core.iterators.SortedKeyValueIterator;
+import org.apache.commons.jexl3.parser.JexlNode;
+import org.apache.hadoop.io.Text;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Preconditions;
+
+import datawave.next.stats.DocumentIteratorStats;
+import datawave.query.data.parsers.FieldIndexKey;
+import datawave.query.jexl.visitors.JexlStringBuildingVisitor;
+
+/**
+ * An abstract document id iterator. Each extending iterator must handle
+ * <ul>
+ * <li>parsing relevant information from the JexlNode</li>
+ * <li>build the correct scan range given context</li>
+ * <li>handle a datatype miss</li>
+ * </ul>
+ */
+public abstract class BaseDocIdIterator implements Iterator<Key> {
+
+    private static final Logger log = LoggerFactory.getLogger(BaseDocIdIterator.class);
+
+    protected final JexlNode node;
+    protected String nodeString;
+
+    protected final String row;
+    protected String prefix;
+    protected Collection<ByteSequence> columnFamilies;
+
+    protected TreeSet<String> datatypeFilter;
+    protected LongRange timeFilter;
+    protected String suffix; // if executing within the context of a document range
+
+    protected String minDatatype;
+    protected String maxDatatype;
+
+    protected String minDatatypeUid;
+    protected String maxDatatypeUid;
+
+    protected String field;
+
+    protected Key tk;
+    protected final SortedKeyValueIterator<Key,Value> source;
+
+    protected boolean seenSeek = false;
+    protected final FieldIndexKey parser = new FieldIndexKey();
+    protected final DocumentIteratorStats stats = new DocumentIteratorStats();
```

Comment:
might it be valuable to be able to disable stats?

Reply:
I plan to make this configurable in the near future